link elements should be able to fire more than one load / error event

rniwa opened the pull request:
#### d19e2bc5b2348544278300e500d500dc3a105df3
<pre>
link elements should be able to fire more than one load / error event
<a href="https://bugs.webkit.org/show_bug.cgi?id=232309">https://bugs.webkit.org/show_bug.cgi?id=232309</a>

Reviewed by NOBODY (OOPS!).

Based on a patch written by Chris Dumez.

This patch makes link element emit more than one load event each time resource is loaded,
and fixes the bug that setting rel content attribute to the same value resulted in the resource to be reloaded.
New behavior matches that of Chrome and Firefox.

* LayoutTests/TestExpectations: Unskip now passing tests.
* LayoutTests/fast/dom/HTMLLinkElement/link-preload-load-once-expected.txt: Added.
* LayoutTests/fast/dom/HTMLLinkElement/link-preload-load-once.html: Added.
* LayoutTests/fast/dom/HTMLLinkElement/link-stylesheet-load-once-expected.txt: Added.
* LayoutTests/fast/dom/HTMLLinkElement/link-stylesheet-load-once.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-multiple-error-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-multiple-load-events-expected.txt:
* LayoutTests/platform/mac-wk1/fast/dom/HTMLLinkElement/link-preload-load-once-expected.txt:
* LayoutTests/platform/mac-wk1/fast/dom/HTMLLinkElement/link-stylesheet-load-once-expected.txt:
* LayoutTests/platform/win/fast/dom/HTMLLinkElement/link-preload-load-once-expected.txt:
* LayoutTests/platform/win/fast/dom/HTMLLinkElement/link-stylesheet-load-once-expected.txt:

* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::HTMLLinkElement):
(WebCore::HTMLLinkElement::parseAttribute):
(WebCore::HTMLLinkElement::notifyLoadedSheetAndAllCriticalSubresources):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/LinkRelAttribute.h:
(WebCore::operator==): Added.
</pre>

cdumez reviewed (approved):
LGTM

darinadler commented on LayoutTests/fast/dom/HTMLLinkElement/link-stylesheet-load-once.html:
```diff
@@ -0,0 +1,32 @@
+function didLoad(element)
+{
+    if (!loadCount) {
+        document.querySelector('link').rel = 'StyleSheet';
```
A few thoughts about how we might want to improve this test to help us further protect interoperability:

1) Make this part of WPT instead of a WebKit-only test.
2) Test whether the resources are reloaded, too, not just whether the load event fires.
3) Test other types of pairs of “same” rel attribute values, such as values that are literally identical, values that have case differences, values that have leading or trailing white space.
4) Test cases other than style sheets.

None of this seems strictly required, but honestly without more testing I am not sure that testing the **parsed** version of the link attribute for equality is precisely the right thing to do. Earlier on in my experience with the web platform I used to assume that in cases where something “changed” to the same effective value doing the minimum work was good for better performance and less flashiness. Later I learned that websites often rely on side effects of setting something to the same value, whether intentionally or not. Thus this is something that may need to be carefully tested and specified since intuition was not a good guide for me.

I’m not sure that all the browsers would agree about when the change to the attribute should trigger a reload, and any differences can lead to interoperability problems later.

rniwa replied:
I think it's a bit tricky to detect how many loads had happened purely in JS. It probably needs to be a HTTP test. For now, I'm going to use `dumpResourceLoadCallbacks` to address (2). I'll add more cases for (3). I'd also a test for "preload" as well for (4) since that triggers load event as well.

rniwa replied:
Tracking this issue in HTML spec as https://github.com/whatwg/html/issues/8138. Without that, there isn't a basis to add WPT.

darinadler commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -167,9 +166,13 @@ void HTMLLinkElement::setDisabledState(bool disabled)
+        m_relAttribute = newRelAttribute;
         if (m_relList)
             m_relList->associatedAttributeValueChanged(value);
+        if (!didMutateRel)
```
Not sure if early return is clearer here than just putting the call to `process()` inside an if statement. Something about the two return statements in a row looks a little inelegant.

rniwa replied:
Yeah, I was thinking of the same so will change.

darinadler commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -167,9 +166,13 @@ void HTMLLinkElement::setDisabledState(bool disabled)
 {
     if (name == relAttr) {
-        m_relAttribute = LinkRelAttribute(document(), value);
+        auto newRelAttribute = LinkRelAttribute(document(), value);
```
I would be tempted to use the word “parsed” in this local variable name. In fact, the class `LinkRelAttribute` annd the data member `m_relAttribute` also should probably should have the word “parsed” in their names, since they do not contain the actual value of the rel attribute! No need to change this, just noticing in passing.

rniwa replied:
I'd change this local variable to `parsedRel`.

darinadler reviewed (approved):
Great fix! Some thoughts about possible follow-up or refinement.

rniwa committed:
link elements should be able to fire more than one load / error event
https://bugs.webkit.org/show_bug.cgi?id=232309

Reviewed by Chris Dumez and Darin Adler.

Based on a patch written by Chris Dumez.

This patch makes link element emit more than one load event each time resource is loaded,
and fixes the bug that setting rel content attribute to the same value resulted in the resource to be reloaded.
New behavior matches that of Chrome and Firefox.

* LayoutTests/TestExpectations: Unskip now passing tests.
* LayoutTests/fast/dom/HTMLLinkElement/link-preload-load-once-expected.txt: Added.
* LayoutTests/fast/dom/HTMLLinkElement/link-preload-load-once.html: Added.
* LayoutTests/fast/dom/HTMLLinkElement/link-stylesheet-load-once-expected.txt: Added.
* LayoutTests/fast/dom/HTMLLinkElement/link-stylesheet-load-once.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-multiple-error-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-multiple-load-events-expected.txt:
* LayoutTests/platform/mac-wk1/fast/dom/HTMLLinkElement/link-preload-load-once-expected.txt:
* LayoutTests/platform/mac-wk1/fast/dom/HTMLLinkElement/link-stylesheet-load-once-expected.txt:
* LayoutTests/platform/win/fast/dom/HTMLLinkElement/link-preload-load-once-expected.txt:
* LayoutTests/platform/win/fast/dom/HTMLLinkElement/link-stylesheet-load-once-expected.txt:

* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::HTMLLinkElement):
(WebCore::HTMLLinkElement::parseAttribute):
(WebCore::HTMLLinkElement::notifyLoadedSheetAndAllCriticalSubresources):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/LinkRelAttribute.h:
(WebCore::operator==): Added.

Canonical link: https://commits.webkit.org/252943@main
